### DISCUSSION TITLE
New workflow for ArmoniK.Api

### DISCUSSION THREAD

**Discussion by esoubiran-aneo** [aneoconsulting/ArmoniK#822] [Ideas]
Hello, 

In order to have a better CI/CD in the [repository ArmoniK.Api](https://github.com/aneoconsulting/ArmoniK.Api), here a proposal to update workflow for CI and when a package is released.

## Proposal

In order to simplify workflow and yml files, here requirements and a workflow.

Requirements
- C# packages must be build on PR, on Release
- Python package must be build on Release
- Angular package must be build on Release
- CI (composition of proto formatting, csharp formatting and build docs) must run on every PR and on push on main

The workflow will be
- Someone create a PR
- CI workflow run and if success, CD for C# run using a random tag
- When PR is merged to main
- CI workflow run and if success, CD for C# run using a random tag
- When a release is created (using GitHub UI (better because you can use the automatic changelog creation) or CLI)
- Every packages are build and released using the release version

With this, no more release branch and dev will be able to be notify on each release. Also, this will allow use to have a changelog using the tag history).

What do you think of this ? 💡

**Comment by aneojgurhem** [comment]
So you drop the existing auto tagging process ?
And the user has to set the tag by himself. How do you ensure the coherence of the tags ?

**Reply by esoubiran-aneo** [reply]
Finally, the question is do you prefer to push to release or to create a release from GitHub. 

I think its important to keep in mind that pushing to release will not create the changelog (without going on GitHub) and it will not notify subscribers because it creates a tag and not a release.

Actually, no notification when a release is created for subscribers is an issue because ArmoniK.Api is a requirement for many others repo.

**Reply by esoubiran-aneo** [reply]
About the tag creation, I think it can be done using a PR setting correct filed in every projects (can be done with a script). Then, when a release is created, it will use tag from project.

**Reply by aneojgurhem** [reply]
Proposal : we can tag every merge on main and make releases only when needed. The GH release would only notify.

**Reply by esoubiran-aneo** [reply]
Seems perfect to me. Will produce new the pipeline soon.

**Reply by aneojgurhem** [reply]
https://github.com/marketplace/actions/github-actions-create-tag

**Reply by esoubiran-aneo** [reply]
Yes but how to use this? You don't want to tag every merge on main.

Here the implementation of the proposal : https://github.com/aneoconsulting/ArmoniK.Api/pull/94

**Reply by Barbapapazes** [reply]
So, final proposal to create a release is:

- Checkout on main
- Create a new branch
- Use a script (on your local machine) to update every version. This script will create a new tag using semver or use your input
- Create a new pull request to main
- An action will run to verify that each project have the same version
- When ready, merge the PR
- Then, create a new release (which create the tag) using GitHub
- Once the tag is created, actions will build and publish packages using the new version


To be sure that all of this work, we need to use conventional commit and we need to write version in projects files. 


This will require to create a `how-to-contribute` file to be sure that people have the workflow in mind.

In parallel, every merge to main will generate a release for C# projects. The proposal is:

- Create the new tag with this pattern `z.y.x-buildId-branchName-commitSha`
- Update each version from every projects with the new tag (only during the action) 
- Build packages
- Push packages to the registry

All of this allow us to simplify the process to build, pack and release a new version from `ArmoniK.Api`. In the same time, this way is the most secure one because (with good settings enabled) only maintainers can merge PR (which allow new versions to be released). Also, many automations are used to be check if tag are correct and everywhere.

**Reply by aneojgurhem** [reply]
Moreover, we can also restrict tag creation to maintainers

**Reply by esoubiran-aneo** [reply]
- [x] Create new release process
- [x] Create update all script
- [x] Create check all script
- [x] Create edge release process
